File: Excel/618 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/618 Project Plan with Relationship.xlsx"
input = read_excel(path, range = "A2:D12")
test  = read_excel(path, range = "F2:G12")
        
input = input %>%
  mutate(
    Start = if_else(Task == "Start", `Plan Start`, as.Date(NA)),
    End = if_else(Task == "Start", `Plan Start` + days(Duration), as.Date(NA))
  )

predecessors = unique(input$Predecessor[!is.na(input$Predecessor)])

for (pred in predecessors) {
  input = input %>%
    mutate(
      Start = case_when(
        Predecessor == pred ~ input$End[input$Task == pred],
        !is.na(Start) ~ Start,
        TRUE ~ as.Date(NA)
      ),
      End = case_when(
        Predecessor == pred ~ Start + days(Duration),
        !is.na(End) ~ End,
        TRUE ~ as.Date(NA)
      )
    )
}
 
all.equal(test, input %>% select(Start, End), check.attributes = FALSE)
# [1] TRUE

File: Excel/600-699/618/618 Challenge.py
import pandas as pd
from datetime import timedelta

path = "618 Project Plan with Relationship.xlsx"
input = pd.read_excel(path, usecols="A:E", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="F:G", skiprows=1, nrows=10)

input['Start'] = input.apply(lambda row: row['Plan Start'] if row['Task'] == 'Start' else pd.NaT, axis=1)
input['End'] = input['Start'] + input['Duration'].apply(lambda d: timedelta(days=d))

predecessors = input['Predecessor '].dropna().unique()

for pred in predecessors:
    pred_end = input.loc[input['Task'] == pred, 'End'].values[0]
    input.loc[input['Predecessor '] == pred, 'Start'] = pred_end
    input['End'] = input['Start'] + input['Duration'].apply(lambda d: timedelta(days=d))

result = input[['Start', 'End']].reset_index(drop=True)
test.columns = result.columns

print(all(result == test)) # True